src/parquet/arrow/reader.cc

```diff
@@ -874,6 +875,134 @@ struct TransferFunctor<
+    *high <<= (length - 8) * CHAR_BIT;
+    *high |= used_high_bits_value;
+
+    // low bytes
+    *low = BytesToInteger<uint64_t>(bytes, length - 8, length);
+  }
+}
+
+/// \brief Convert an array of FixedLenByteArrays to an arrow::DecimalArray
+/// We do this by:
+/// 1. Creating a arrow::FixedSizeBinaryArray from the RecordReader's builder
+/// 2. Allocating a buffer for the arrow::DecimalArray
+/// 3. Converting the big-endian bytes in the FixedSizeBinaryArray to two integers
+///    representing the high and low bits of each decimal value.
+template <>
+struct TransferFunctor<::arrow::DecimalType, FLBAType> {
+  Status operator()(RecordReader* reader, MemoryPool* pool,
+                    const std::shared_ptr<::arrow::DataType>& type,
+                    std::shared_ptr<Array>* out) {
+    DCHECK_EQ(type->id(), ::arrow::Type::DECIMAL);
+
+    // Finish the built data into a temporary array
+    std::shared_ptr<Array> array;
+    RETURN_NOT_OK(reader->builder()->Finish(&array));
+    const auto& fixed_size_binary_array =
+        static_cast<const ::arrow::FixedSizeBinaryArray&>(*array);
+
+    // Get the byte width of the values in the FixedSizeBinaryArray. Most of the time
+    // this will be different from the decimal array width because we write the minimum
+    // number of bytes necessary to represent a given precision
+    const int32_t byte_width =
+        static_cast<const ::arrow::FixedSizeBinaryType&>(*fixed_size_binary_array.type())
+            .byte_width();
+
+    // The byte width of each decimal value
+    const int32_t type_length =
+        static_cast<const ::arrow::DecimalType&>(*type).byte_width();
+
+    // number of elements in the entire array
+    const int64_t length = fixed_size_binary_array.length();
+
+    // allocate memory for the decimal array
+    std::shared_ptr<Buffer> data;
+    RETURN_NOT_OK(::arrow::AllocateBuffer(pool, length * type_length, &data));
+
+    // raw bytes that we can write to
+    uint8_t* out_ptr = data->mutable_data();
+
+    auto raw_bytes_to_decimal_bytes = [byte_width](const uint8_t* value,
+                                                   uint8_t* out_buf) {
+      // view the first 8 bytes as a signed 64-bit integer
+      auto high = reinterpret_cast<int64_t*>(out_buf);
+
+      // view the second 8 bytes as an unsigned 64-bit integer
+      auto low = reinterpret_cast<uint64_t*>(out_buf + sizeof(int64_t));
+
+      // Convert the fixed size binary array bytes into a Decimal128 compatible layout
+      BytesToIntegerPair(value, byte_width, high, low);
+    };
+
+    // convert each FixedSizeBinary value to valid decimal bytes
+    const int64_t null_count = fixed_size_binary_array.null_count();
+    if (null_count > 0) {
+      for (int64_t i = 0; i < length; ++i, out_ptr += type_length) {
+        if (!fixed_size_binary_array.IsNull(i)) {
+          raw_bytes_to_decimal_bytes(fixed_size_binary_array.GetValue(i), out_ptr);
```

Comment:
The new buffer will have defined slots because I take a view on the bytes (`int64_t*` for high bits, `uint64_t*` for low bits) and assign values `*high`/`*low`. That happens in `BytesToIntegerPair`.